src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3503,6 +3503,20 @@ internal override bool CompileMethods(
                 }
 
                 SynthesizedMetadataCompiler.ProcessSynthesizedMembers(this, moduleBeingBuilt, cancellationToken);
+
+                if (moduleBeingBuilt.OutputKind.IsApplication())
+                {
+                    var entryPoint = GetEntryPointAndDiagnostics(cancellationToken);
+                    diagnostics.AddRange(entryPoint.Diagnostics.Diagnostics);
+                    if (entryPoint.MethodSymbol != null)
+                    {
+                        moduleBeingBuilt.SetPEEntryPoint(entryPoint.MethodSymbol, diagnostics);
```

Comment:
>moduleBeingBuilt.SetPEEntryPoint(entryPoint.MethodSymbol, diagnostics); [](http://example.com/codeflow?start=24&length=71)

Is this going to do a right thing for an `async` Main? It looks like ```GetEntryPoint``` helper synthesizes a wrapper method as an entry point in that case.

Reply:
It will set the async Main as the entrypoint, not the generated one. I think that's fine for ref assembly.